rest_api/aggregation/server_side/expressions.py

```diff
@@ -10,80 +10,90 @@ def resolve(query, element):
 
 
 class Expression:
-    default_return_value = None
-
-    def __init__(self, *args):
+    def __init__(self, *args, filter_func=None):
         self.args = args
+        self.filter_func = filter_func
+
+    def evaluate(self, e):
+        raise NotImplementedError
+
+
+class Calculation(Expression):
+    default_return_value = None
 
     def _expression(self, *args):
-        return 0
+        raise NotImplementedError
 
-    def _evaluate(self, *args):
-        try:
-            return self._expression(*args)
-        except (ZeroDivisionError, TypeError):
-            return self.default_return_value
+    def _resolve_element(self, element, query_string):
+        e = element.copy()
+        queries = query_string.split('.')
 
-    def _resolve_element(self, e, param):
-        pass
+        for q in queries[:-1]:
+            e = e.get(q, {})
+        return e.get(queries[-1])
 
     def evaluate(self, e):
-        data = []
         if e is None:
             return None
+
+        _args = []
         for a in self.args:
             if isinstance(a, Expression):
                 data_point = a.evaluate(e)
             else:
                 data_point = self._resolve_element(e, a)
+
             if data_point is None:
                 return self.default_return_value
-            data.append(data_point)
-        return self._evaluate(*data)
 
+            _args.append(data_point)
 
-class SingleExp(Expression):
-    def _resolve_element(self, element, param):
-        pparam = param.split('.')
-        for e in pparam[:-1]:
-            element = element.get(e)
-        return element.get(pparam[-1])
+        try:
+            return self._expression(*_args)
+        except (ZeroDivisionError, TypeError):
+            return self.default_return_value
 
 
-class Accumulation(Expression):
-    def __init__(self, *args, filter_func=None):
-        super().__init__(*args)
-        self.filter_func = filter_func
+class Accumulation(Calculation):
+    def _resolve_element(self, element, query_string):
+        e = element.copy()
+        queries = query_string.split('.')
+        subelements = e.get(queries[0])
+        if subelements is None:
+            return None
 
-    def _resolve_element(self, element, param):
-        pparam = param.split('.')
-        for p in pparam[:-1]:
-            element = element.get(p)
-        if element is None:
-            return
+        assert type(subelements) is list, 'In %s: element %s is not a list' % (self.__class__.__name__, element)
```

Comment:
This changes the behaviour of `Accumulation` to assume the first element of the `query_string` returns a `list`. 
Before `Accumulation`  was assuming that the one before last was returning a list. 
Was this intended? 
If yes could this be made clearer in a comment?

Reply:
This is a behaviour change, but not one that affects what is currently there because we've only ever drilled down by one item. This was originally here because new things like `coverage_at_15X` drill down multiple times. Ideally the behaviour would be controllable - will add a docstring/todo.